### PR TITLE
Allow metrics related threads and mbeans cleanup on sensei server shutdown

### PR DESCRIPTION
**Pull request by chtyim**
Original, the metrics related threads and mbeans are cleanup by shutdown hook. The major changes is to replaces usage of Metrics.xxx() to MetricFactory.xxx(), which the MetricFactory provides static start and stop methods that will be called when sensei server start/stop accordingly, allowing proper resources cleanup.

### CODE REVIEW

File: sensei-core/src/main/java/com/senseidb/indexing/DefaultStreamingIndexingManager.java [hunk 1]
```diff
@@ -219,6 +216,10 @@ public void start() throws Exception {
 		  logger.warn("no data stream configured, no indexing events are flowing.");
 		}
 		else{
+      _providerBatchSizeMeter = registerMeter("provider-batch-size", "provide-batch-size");
+      _updateBatchSizeMeter = registerMeter("update-batch-size", "update-batch-size");
+      _eventMeter = registerMeter("indexing-events", "indexing-events");
```
Comment (santip) [inline sensei-core/src/main/java/com/senseidb/indexing/DefaultStreamingIndexingManager.java:221]: Looks like you're no longer skipping over errors, and if registering a meter fails, startup will fail. This is not the previous behavior, are we sure this won't break anyone?

File: sensei-core/src/main/java/com/senseidb/indexing/TimeBasedIndexSelector.java [hunk 2]
```diff
@@ -38,16 +38,10 @@
...
   private String facetName;
-  private static Counter processedReadersCount;
-  private static Counter filteredReadersCount;
-  static{
-    // register jmx monitoring for timers
-      MetricName processedReadersMetric = new MetricName(MetricsConstants.Domain, "timeBasedIndexPruner","processedReaderCount");
-      processedReadersCount = Metrics.newCounter(processedReadersMetric);
-      MetricName filteredReadersMetric = new MetricName(MetricsConstants.Domain,"timeBasedIndexPruner","filteredReaderCount");
-      filteredReadersCount = Metrics.newCounter(filteredReadersMetric);
-
```
Comment (santip) [inline sensei-core/src/main/java/com/senseidb/indexing/TimeBasedIndexSelector.java:49]: here it looks like failed metrics would halt the class loading, so perhaps that's not an issue

File: sensei-core/src/main/java/com/senseidb/indexing/activity/CompositeActivityValues.java [hunk 3]
```diff
@@ -79,21 +79,21 @@
...
+  protected final Counter insertedDocumentsCounter;
+  protected final Counter totalUpdatesCounter;
+
+  CompositeActivityValues() {
+    reclaimedDocumentsCounter = MetricFactory.newCounter(new MetricName(CompositeActivityValues.class,
+                                                                        "reclaimedActivityDocs"));
+    currentDocumentsCounter = MetricFactory.newCounter(new MetricName(CompositeActivityValues.class, "currentActivityDocs"));
+    deletedDocumentsCounter = MetricFactory.newCounter(new MetricName(CompositeActivityValues.class, "deletedActivityDocs"));
+    insertedDocumentsCounter = MetricFactory.newCounter(new MetricName(CompositeActivityValues.class, "insertedActivityDocs"));
+    totalUpdatesCounter = MetricFactory.newCounter(new MetricName(CompositeActivityValues.class, "totalUpdatesCounter"));
```
Comment (santip) [inline sensei-core/src/main/java/com/senseidb/indexing/activity/CompositeActivityValues.java:94]: Why build the counters on Constructor and not init()?

**Comment by santip** [comment]
In general, looks good to me.

File: sensei-core/src/main/java/com/senseidb/indexing/DefaultStreamingIndexingManager.java (hunk 1 above)
Comment (chtyim) [inline sensei-core/src/main/java/com/senseidb/indexing/DefaultStreamingIndexingManager.java:221]: It passes all tests. Besides, in the old code, it left some meters being null when there is setup failure, which will cause NPE eventually. I think it's better to fail earlier than throwing NPE some time later depending on the execution path.

**Pull request merged**